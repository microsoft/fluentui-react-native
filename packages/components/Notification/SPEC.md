# Notification

## Variants

| Variant           | Type  | Light     | Dark      |
| ----------------- | ----- | --------- | --------- |
| Primary           | Toast | `#EBF3FC` | `#082338` |
| Neutral           | Toast | `#FAFAFA` | `#333333` |
| PrimaryBar        | Bar   | `#EBF3FC` | `#082338` |
| PrimaryOutlineBar | Bar   | `#FFFFFF` | `#000000` |
| NeutralBar        | Bar   | `#F0F0F0` | `#3D3D3D` |
| Danger            | Toast | `#FDF6F6` | `#3F1011` |
| Warning           | Toast | `#FFFBD6` | `#4C4400` |

## Icon

Toast notifications can include an optional `icon` that appears to the left of the `message`.

## Title

Toast notifications can include an optional `title` that appears on top of the `message`. The `message` will then become subtext.

## Button

Depending on what props are passed in, either an action button or dismiss button will be rendered as the `button`.

- If the props contain an action to be performed (an onActionPress callback function) and an action text, the action button will be rendered.
- If the props contain an action to be performed but no action text, the dismiss button will be rendered.
- In all other cases, no button will be rendered.

## Callback Functions

| Prop            | Description                                           |
| --------------- | ----------------------------------------------------- |
| `onPress`       | Action performed by pressing the entire notification. |
| `onActionPress` | Action performed by pressing the `button`.            |
