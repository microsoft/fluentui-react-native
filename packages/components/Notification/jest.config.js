const { configureReactNativeJest } = require('@fluentui-react-native/scripts');
