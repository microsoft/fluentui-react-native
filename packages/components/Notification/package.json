{
  "name": "@fluentui-react-native/notification",
  "version": "0.24.18",
  "description": "add component-description",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "clean": "fluentui-scripts clean",
    "depcheck": "fluentui-scripts depcheck",
    "just": "fluentui-scripts",
    "lint": "fluentui-scripts eslint",
    "test": "fluentui-scripts jest",
    "update-snapshots": "fluentui-scripts jest -u",
    "prettier": "fluentui-scripts prettier",
    "prettier-fix": "fluentui-scripts prettier --fix true"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/fluentui-react-native.git",
    "directory": "packages/components/Notification"
  },
  "dependencies": {
    "@fluentui-react-native/adapters": "workspace:*",
    "@fluentui-react-native/button": "workspace:*",
    "@fluentui-react-native/experimental-appearance-additions": "workspace:*",
    "@fluentui-react-native/experimental-shadow": "workspace:*",
    "@fluentui-react-native/framework": "workspace:*",
    "@fluentui-react-native/icon": "workspace:*",
    "@fluentui-react-native/interactive-hooks": "workspace:*",
    "@fluentui-react-native/pressable": "workspace:*",
    "@fluentui-react-native/text": "workspace:*",
    "@fluentui-react-native/theme-tokens": "workspace:*",
    "@fluentui-react-native/theme-types": "workspace:*",
    "@fluentui-react-native/tokens": "workspace:*",
    "@fluentui-react-native/use-styling": "workspace:*"
  },
  "devDependencies": {
    "@fluentui-react-native/eslint-config-rules": "workspace:*",
    "@fluentui-react-native/scripts": "workspace:*",
    "@fluentui-react-native/test-tools": "workspace:*",
    "@office-iss/react-native-win32": "^0.73.0",
    "@react-native/babel-preset": "^0.73.0",
    "@react-native/metro-config": "^0.73.0",
    "react": "18.2.0",
    "react-native": "^0.73.0",
    "react-native-svg": "^14.0.0"
  },
  "peerDependencies": {
    "@office-iss/react-native-win32": "^0.73.0",
    "react": "18.2.0",
    "react-native": "^0.73.0",
    "react-native-macos": "^0.73.0",
    "react-native-svg": "^14.0.0",
    "react-native-windows": "^0.73.0"
  },
  "author": "",
  "license": "MIT",
  "rnx-kit": {
    "kitType": "library",
    "alignDeps": {
      "presets": [
        "microsoft/react-native"
      ],
      "requirements": [
        "react-native@0.73"
      ],
      "capabilities": [
        "core",
        "core-android",
        "core-ios",
        "react",
        "svg"
      ]
    }
  },
  "peerDependenciesMeta": {
    "@office-iss/react-native-win32": {
      "optional": true
    },
    "react-native-macos": {
      "optional": true
    },
    "react-native-windows": {
      "optional": true
    }
  }
}
