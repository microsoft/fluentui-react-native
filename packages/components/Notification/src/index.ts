export * from './Notification';
export * from './Notification.types';
