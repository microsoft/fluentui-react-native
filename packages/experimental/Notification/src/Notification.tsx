import React, {Component} from 'react';
import {Text} from 'react-native';


export default class Notification extends Component<any> {
  render() {
    return (
      <Text>Notification!</Text>
    );
  }
}