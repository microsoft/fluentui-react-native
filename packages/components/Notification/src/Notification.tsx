/** @jsx withSlots */
import { notification, NotificationType, NotificationProps } from './Notification.types';
import { Pressable } from '@fluentui-react-native/pressable';
import { Platform, PressableProps, useWindowDimensions, View, ViewStyle } from 'react-native';
import { Icon } from '@fluentui-react-native/icon';
import { TextV1 as Text } from '@fluentui-react-native/text';
import { stylingSettings } from './Notification.styling';
import { compose, mergeProps, withSlots, UseSlots } from '@fluentui-react-native/framework';
import { useMemo } from 'react';
import { createIconProps } from '@fluentui-react-native/interactive-hooks';
import { NotificationButton, createNotificationButtonProps } from './Notification.helper';
import { Shadow } from '@fluentui-react-native/experimental-shadow';

type SizeClassIOS = 'regular' | 'compact' | undefined;

/**
 * Hook that "guesses" our Size Class on iOS based on our window width.
 * Note: this is hacky and should not be used.
 *
 * For more information about Size Classes, see the following:
 * https://developer.apple.com/documentation/uikit/uitraitcollection
 * https://developer.apple.com/design/human-interface-guidelines/foundations/layout/#platform-considerations
 * @returns SizeClassIOS: enum determining our size class
 */
const useSizeClassIOS_DO_NOT_USE: () => SizeClassIOS = () => {
  const width = useWindowDimensions().width;
  if (Platform.OS === 'ios') {
    if (Platform.isPad && width > 375) {
      return 'regular';
    } else {
      return 'compact';
    }
  } else {
    return undefined;
  }
};
/**
 * A function which determines if a set of styles should be applied to the component given the current state and props of the Notification.
 *
 * @param layer The name of the state that is being checked for
 * @param userProps The props that were passed into the Notification
 * @returns Whether the styles that are assigned to the layer should be applied to the Notification
 */
export const notificationLookup = (layer: string, userProps: NotificationProps): boolean => {
  return (
    layer === userProps.variant ||
    (layer === 'hasTitle' && userProps.title != undefined) ||
    (layer === 'isBar' && ['primaryOutlineBar', 'primaryBar', 'neutralBar'].includes(userProps.variant))
  );
};

export const Notification = compose<NotificationType>({
  displayName: notification,
  ...stylingSettings,
  slots: {
    root: Pressable,
    icon: Icon,
    contentContainer: View,
    title: Text,
    message: Text,
    action: NotificationButton,
    shadow: Shadow,
  },
  useRender: (userProps: NotificationProps, useSlots: UseSlots<NotificationType>) => {
    const Slots = useSlots(userProps, (layer) => notificationLookup(layer, userProps));
    const isBar = ['primaryOutlineBar', 'primaryBar', 'neutralBar'].includes(userProps.variant);
    const width = useWindowDimensions().width / 2;
    const sizeClass = useSizeClassIOS_DO_NOT_USE();
    const onActionPress = userProps.onActionPress;

    const rootStyle: ViewStyle = useMemo(() => {
      const marginHorizontal = isBar ? 0 : 16;
      if (sizeClass === 'regular' && !isBar) {
        return { alignSelf: 'center', marginHorizontal: marginHorizontal, width: width };
      } else {
        return { marginHorizontal: marginHorizontal };
      }
    }, [isBar, width]);
    const messageStyle: ViewStyle = useMemo(() => {
      const alignSelf = onActionPress ? 'flex-start' : 'center';
      return { alignSelf: alignSelf };
    }, [onActionPress]);

    return (final: NotificationProps, ...children: React.ReactNode[]) => {
      const { variant, icon, title, action, onActionPress, ...rest } = mergeProps(userProps, final);
      const mergedProps = mergeProps<PressableProps>(rest, rootStyle);
      const iconProps = createIconProps(icon);
      const notificationButtonProps = createNotificationButtonProps(userProps);

      return (
        <Slots.shadow>
          <Slots.root {...mergedProps}>
            {icon && <Slots.icon {...iconProps} accessible={false} />}
            <Slots.contentContainer accessible={true}>
              {title && <Slots.title>{title}</Slots.title>}
              <Slots.message style={messageStyle}>{children}</Slots.message>
            </Slots.contentContainer>
            {onActionPress && <Slots.action {...notificationButtonProps} />}
          </Slots.root>
        </Slots.shadow>
      );
    };
  },
});
