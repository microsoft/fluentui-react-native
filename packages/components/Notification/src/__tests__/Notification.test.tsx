import * as React from 'react';

import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import { Notification } from '../Notification';

describe('Notification component tests', () => {
  beforeAll(() => {
    jest.mock('@fluentui-react-native/experimental-appearance-additions', () => ({
      useHorizontalSizeClass: 'regular',
    }));
  });

  it('Notification default', () => {
    const tree = renderer
      .create(
        <Notification
          variant={'primary'}
          action="Undo"
          onPress={() => {
            console.log('Notification tapped');
          }}
          onActionPress={() => {
            console.log('Undo tapped');
          }}
        >
          Mail Archived
        </Notification>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Notification simple rendering does not invalidate styling', () => {
    checkRenderConsistency(
      () => (
        <Notification
          variant={'primary'}
          action="Undo"
          onPress={() => {
            console.log('Notification tapped');
          }}
          onActionPress={() => {
            console.log('Undo tapped');
          }}
        >
          Mail Archived
        </Notification>
      ),
      2,
    );
  });

  it('Notification re-renders correctly', () => {
    checkReRender(
      () => (
        <Notification
          variant={'primary'}
          action="Undo"
          onPress={() => {
            console.log('Notification tapped');
          }}
          onActionPress={() => {
            console.log('Undo tapped');
          }}
        >
          Mail Archived
        </Notification>
      ),
      2,
    );
  });
});
