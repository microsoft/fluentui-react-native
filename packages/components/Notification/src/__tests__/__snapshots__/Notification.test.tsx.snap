// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component tests Notification default 1`] = `
<View
  style={
    Object {
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 8,
        "width": 0,
      },
      "shadowOpacity": 0.14,
      "shadowRadius": 8,
    }
  }
>
  <View
    style={
      Object {
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0.12,
        "shadowRadius": 1,
      }
    }
  >
    <View
      marginHorizontal={16}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onPress={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#EBF3FC",
          "borderColor": "transparent",
          "borderRadius": 12,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "minHeight": 64,
          "padding": 16,
        }
      }
    >
      <View
        accessible={true}
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={0}
          style={
            Object {
              "alignSelf": "flex-start",
              "color": "#0F6CBD",
              "fontFamily": "Segoe UI",
              "fontSize": 15,
              "fontWeight": "600",
              "margin": 0,
            }
          }
        >
          Mail Archived
        </Text>
      </View>
      <View
        accessibilityLabel="Undo"
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        appearance="subtle"
        color="#0F6CBD"
        disabled={false}
        disabledColor="#2886DE"
        enableFocusRing={true}
        focusable={true}
        minWidth={0}
        onAccessibilityTap={[Function]}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onPress={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        padding={0}
        paddingHorizontal={0}
        pressedColor="#0F6CBD30"
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "transparent",
            "borderColor": "#ffffff",
            "borderRadius": 4,
            "borderWidth": 1,
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "center",
            "marginStart": 16,
            "minWidth": 96,
            "padding": 5,
            "paddingHorizontal": 11,
          }
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={0}
          style={
            Object {
              "color": "#0F6CBD",
              "fontFamily": "Segoe UI",
              "fontSize": 15,
              "fontWeight": "600",
              "margin": 0,
              "marginBottom": 0,
              "marginEnd": 0,
              "marginStart": 0,
              "marginTop": 0,
            }
          }
        >
          Undo
        </Text>
      </View>
    </View>
  </View>
</View>
`;
